File: src/graalvm/micronaut/src/main/java/com/benchmark/controller/CacheController.java
package com.benchmark.controller;

import com.benchmark.service.CacheService;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.QueryValue;

import jakarta.inject.Inject;
import java.time.Instant;
import java.util.Map;

@Controller
public class CacheController {

    // The TTL is part of the response contract and must match what is written
    // to Redis. See contracts/rest/canonical-payloads.md.
    private static final int CACHE_TTL_SECONDS = 300;
    private final CacheService cacheService;

    @Inject
    public CacheController(CacheService cacheService) {
        this.cacheService = cacheService;
    }

    @Get("/cache")
    public Map<String, Object> cache(@QueryValue(defaultValue = "test") String key) {
        CacheService.CacheResult result = cacheService.getOrSet(key);

        return Map.of(
            "key", result.getKey(),
            "value", result.getValue(),
            "cached", result.isCached(),
            "ttl", CACHE_TTL_SECONDS,
            "timestamp", Instant.now().toString()
        );
    }
}

File: src/csharp/Controllers/Controllers/CacheController.cs
using BenchmarkControllers.Services;
using Microsoft.AspNetCore.Mvc;

namespace BenchmarkControllers.Controllers;

[ApiController]
public class CacheController : ControllerBase
{
    // The TTL is part of the response contract and must match the expiry the
    // cache service writes. See contracts/rest/canonical-payloads.md.
    private const int CacheTtlSeconds = 300;

    private readonly ICacheService _cacheService;
    private readonly ILogger<CacheController> _logger;

    public CacheController(ICacheService cacheService, ILogger<CacheController> logger)
    {
        _cacheService = cacheService;
        _logger = logger;
    }

    [HttpGet("/cache")]
    public async Task<IActionResult> GetCache([FromQuery] string? key)
    {
        if (string.IsNullOrEmpty(key))
        {
            return BadRequest(new { error = "Key parameter is required" });
        }

        _logger.LogInformation("Cache request for key: {Key}", key);

        var (value, cached) = await _cacheService.GetOrSetAsync(key, async () =>
        {
            return $"Cached value for {key} at {DateTime.UtcNow:O}";
        });

        return Ok(new
        {
            key,
            value,
            cached,
            ttl = CacheTtlSeconds,
            timestamp = DateTime.UtcNow
        });
    }
}
